lib/auth/webauthn/login.go

```diff
@@ -32,84 +32,67 @@ import (
-// It exists to better scope LoginFlow's use of Identity and to facilitate
-// testing.
-type LoginIdentity interface {
-	userIDStorage
+// loginIdentity contains the subset of services.Identity methods used by
+// loginFlow.
+type loginIdentity interface {
+	GetWebauthnLocalAuth(ctx context.Context, user string) (*types.WebauthnLocalAuth, error) // MFA
+	GetTeleportUserByWebauthnID(ctx context.Context, webID []byte) (string, error)           // Passwordless
 
 	GetMFADevices(ctx context.Context, user string, withSecrets bool) ([]*types.MFADevice, error)
 	UpsertMFADevice(ctx context.Context, user string, d *types.MFADevice) error
-	UpsertWebauthnSessionData(ctx context.Context, user, sessionID string, sd *wantypes.SessionData) error
-	GetWebauthnSessionData(ctx context.Context, user, sessionID string) (*wantypes.SessionData, error)
-	DeleteWebauthnSessionData(ctx context.Context, user, sessionID string) error
-}
-
-// WithDevices returns a LoginIdentity backed by a fixed set of devices.
-// The supplied devices are returned in all GetMFADevices calls.
-func WithDevices(identity LoginIdentity, devs []*types.MFADevice) LoginIdentity {
-	return &loginWithDevices{
-		LoginIdentity: identity,
-		devices:       devs,
-	}
 }
 
-type loginWithDevices struct {
-	LoginIdentity
-	devices []*types.MFADevice
+// sessionIdentity abstracts operations over SessionData storage.
+// * MFA uses per-user variants
+// (services.Identity.Update/Get/DeleteWebauthnSessionData methods).
+// * Passwordless uses global variants
+// (services.Identity.Update/Get/DeleteGlobalWebauthnSessionData methods).
+type sessionIdentity interface {
+	Upsert(ctx context.Context, user string, sd *wantypes.SessionData) error
+	Get(ctx context.Context, user string, challenge string) (*wantypes.SessionData, error)
+	Delete(ctx context.Context, user string, challenge string) error
 }
 
-func (l *loginWithDevices) GetMFADevices(ctx context.Context, user string, withSecrets bool) ([]*types.MFADevice, error) {
-	return l.devices, nil
+// loginFlow implements both MFA and Passwordless authentication, exposing an
+// interface that is the union of both login methods.
+type loginFlow struct {
+	U2F         *types.U2F
+	Webauthn    *types.Webauthn
+	identity    loginIdentity
+	sessionData sessionIdentity
 }
 
-// LoginFlow represents the WebAuthn login procedure (aka authentication).
-//
-// The login flow consists of:
-//
-// 1. Client requests a CredentialAssertion (containing, among other info, a
-//    challenge to be signed)
-// 2. Server runs Begin(), generates a credential assertion.
-// 3. Client validates the assertion, performs a user presence test (usually by
-//    asking the user to touch a secure token), and replies with
-//    CredentialAssertionResponse (containing the signed challenge)
-// 4. Server runs Finish()
-// 5. If all server-side checks are successful, then login/authentication is
-//    complete.
-type LoginFlow struct {
```

Comment:
Note to reviewers: moved to login_mfa.go, along with the other public types/methods that used to be here.